File: exercises/js/day2-step2-issue-token.js
// ============================================================
// Day 2 - Step 2 Exercise: Issue a Custom Token
// Learn: How to create a trust line and issue tokens on XRPL.
// The holder must trust the issuer before receiving tokens.
// Reference materials:
// - https://learn.xrpl-commons.org/course/code-with-the-xrpl/lesson/create-trustline-and-send-currency/
// - https://xrpl.org/docs/references/protocol/transactions/types/trustset
// - https://xrpl.org/docs/references/protocol/transactions/types/payment
// - https://js.xrpl.org/
// ============================================================

const xrpl = require('xrpl');
const fs = require('fs');
const { notifyProgress } = require('../../scripts/notify');

const TESTNET_URL = 'wss://s.altnet.rippletest.net:51233';

async function main() {
  const { wallets } = JSON.parse(fs.readFileSync('../../wallets.json', 'utf-8'));
  const issuerWallet = xrpl.Wallet.fromSeed(wallets[0].seed);
  const holderWallet = xrpl.Wallet.fromSeed(wallets[1].seed);

  // Replace with your initials (3 characters, e.g. 'ACD')
  const currencyCode = 'YOUR_INITIALS';

  const client = new xrpl.Client(TESTNET_URL);
  await client.connect();
  console.log('Connected to XRPL Testnet');

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 1: Create a TrustSet transaction (holder trusts issuer)    ║
  // ║                                                                  ║
  // ║ HINT: {                                                          ║
  // ║   TransactionType: 'TrustSet',                                   ║
  // ║   Account: holderWallet.address,                                 ║
  // ║   LimitAmount: {                                                 ║
  // ║     currency: currencyCode,                                      ║
  // ║     issuer: issuerWallet.address,                                ║
  // ║     value: '1000000'                                             ║
  // ║   }                                                              ║
  // ║ }                                                                ║
  // ║                                                                  ║
  // ║ Submit: client.submitAndWait(tx, { wallet: holderWallet })       ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  console.log('Trust line created!');

  // ╔══════════════════════════════════════════════════════════════════╗
  // ║ TODO 2: Issue tokens (issuer sends tokens to holder)            ║
  // ║                                                                  ║
  // ║ HINT: {                                                          ║
  // ║   TransactionType: 'Payment',                                    ║
  // ║   Account: issuerWallet.address,                                 ║
  // ║   Destination: holderWallet.address,                             ║
  // ║   Amount: {                                                      ║
  // ║     currency: currencyCode,                                      ║
  // ║     issuer: issuerWallet.address,                                ║
  // ║     value: '500'                                                 ║
  // ║   }                                                              ║
  // ║ }                                                                ║
  // ║                                                                  ║
  // ║ Submit: client.submitAndWait(tx, { wallet: issuerWallet })       ║
  // ╚══════════════════════════════════════════════════════════════════╝
  // YOUR CODE HERE

  console.log('Tokens issued!');
  console.log(`\nExplorer: https://testnet.xrpl.org/accounts/${issuerWallet.address}`);
  await notifyProgress('Day 2 Step 2: Issue Token');
  await client.disconnect();
}

main().catch(console.error);

File: exercises/python/day2_step2_issue_token.py
"""
Day 2 Step 2 Exercise: Issue Token
====================================
Create a TrustLine and issue a custom token on XRPL Testnet.

Reference materials:
- https://learn.xrpl-commons.org/course/code-with-the-xrpl/lesson/create-trustline-and-send-currency/
- https://xrpl.org/docs/references/protocol/transactions/types/trustset
- https://xrpl.org/docs/references/protocol/transactions/types/payment
- https://xrpl-py.readthedocs.io/en/stable/
"""

import json
from xrpl.clients import WebsocketClient
from xrpl.wallet import Wallet
from xrpl.models.transactions import TrustSet, Payment
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.transaction import submit_and_wait
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', 'scripts'))
from notify_progress import notify_progress

TESTNET_URL = 'wss://s.altnet.rippletest.net:51233'

# Load wallets
with open('../../wallets.json') as f:
    data = json.load(f)
    wallets = data['wallets']

issuer = Wallet.from_seed(wallets[0]['seed'])
holder = Wallet.from_seed(wallets[1]['seed'])
currency_code = 'TST'

print(f"Issuer: {issuer.address}")
print(f"Holder: {holder.address}")

with WebsocketClient(TESTNET_URL) as client:

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 1: Create a TrustSet from holder to issuer                 ║
    # ║                                                                  ║
    # ║ HINT: TrustSet(account=holder.address,                           ║
    # ║       limit_amount=IssuedCurrencyAmount(                         ║
    # ║           currency=currency_code,                                ║
    # ║           issuer=issuer.address, value='1000000'))                ║
    # ╚══════════════════════════════════════════════════════════════════╝
    trust_tx = None  # Replace with TrustSet(...)
    # trust_result = submit_and_wait(trust_tx, client, holder)

    # ╔══════════════════════════════════════════════════════════════════╗
    # ║ TODO 2: Issue tokens from issuer to holder                      ║
    # ║                                                                  ║
    # ║ HINT: Payment(account=issuer.address,                            ║
    # ║       destination=holder.address,                                ║
    # ║       amount=IssuedCurrencyAmount(currency=currency_code,        ║
    # ║           issuer=issuer.address, value='500'))                    ║
    # ╚══════════════════════════════════════════════════════════════════╝
    payment_tx = None  # Replace with Payment(...)
    # payment_result = submit_and_wait(payment_tx, client, issuer)

# Explorer: https://testnet.xrpl.org/accounts/{address}
print(f"\nExplorer (issuer): https://testnet.xrpl.org/accounts/{issuer.address}")
print(f"Explorer (holder): https://testnet.xrpl.org/accounts/{holder.address}")
notify_progress('Day 2 Step 2: Issue Token')
